### PR TITLE
[SYCL][thinLTO] Seperate module properties and symbol table generation into IR-based analysis

### PR DESCRIPTION
**Pull request by sarnex**
Based on feedback from https://github.com/intel/llvm/pull/14197, I seperated out the code that generates the module properties and symbol table into separate functions that can be called by anyone, and just looks at the IR and entry points.

For now, we still call it inside `sycl-post-link` because we still support the old offloading model, but once we drop support for that we can drop this responsibility from sycl-post-link and only compute it inside `clang-linker-wrapper`, both for normal compilation and thinLTO. 

In a (hopefully soon) future PR I plan to call these functions from `clang-linker-wrapper` when compiling for thinLTO, which we need because we will split early.

Most of this change should be NFC(I). The expected changes are:

1) New option to sycl-post-link to generate the properties file
2) Driver change to NOT pass the option from 1) in thinLTO mode
3) Two minor chages in logic from properties generation, I've called these out inline.

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+  }
+  bool SeenESIMDFunction = false;
+  bool SeenSYCLFunction = false;
+  for (const auto &F : M) {
+    if (llvm::module_split::isESIMDFunction(F))
+      SeenESIMDFunction = true;
+    else if (utils::isSYCLExternalFunction(&F) &&
+             !F.getName().starts_with("__itt"))
+      SeenSYCLFunction = true;
+  }
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp:262]: This logic is new. Previously we used `MD.isESIMD()`

File: llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+  if (SeenESIMDFunction && !SeenSYCLFunction)
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp:263]: This logic is new. Previously we used `MD.isESIMD()`

File: llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+  // GPU backend has a problem with accepting optimization level options in form
+  // described by Level Zero specification (-ze-opt-level=1) when 'invoke_simd'
+  // functionality is involved. JIT compilation results in the following error:
+  //     error: VLD: Failed to compile SPIR-V with following error:
+  //     invalid api option: -ze-opt-level=O1
+  //     -11 (PI_ERROR_BUILD_PROGRAM_FAILURE)
+  // 'if' below essentially preserves the behavior (presumably mistakenly)
+  // implemented in intel/llvm#8763: ignore 'optLevel' property for images which
+  // were produced my merge after ESIMD split
+  if (!SeenESIMDFunction || !SeenSYCLFunction) {
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp:308]: This logic is new. Previously we used `MD.getEntryPointGroup().Props.HasESIMD !=
      module_split::SyclEsimdSplitStatus::SYCL_AND_ESIMD`

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 4]
```diff
@@ -201,6 +196,10 @@ cl::opt<module_split::IRSplitMode> SplitMode(
 cl::opt<bool> DoSymGen{"symbols", cl::desc("generate exported symbol files"),
                        cl::cat(PostLinkCat)};
 
+cl::opt<bool> SkipPropsGen{"skip-properties-gen",
+                           cl::desc("do not generate module properties files"),
+                           cl::cat(PostLinkCat)};
+
```
Comment (bader) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:202]: I suggest we invert the logic and add `properties` option (similar to `symbols`) which will enable generation of properties.

File: llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp [hunk 5]
```diff
@@ -0,0 +1,380 @@
+#include "llvm/SYCLLowerIR/ComputeModuleProperties.h"
```
Comment (bader) [inline llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp:1]: Copyright header is missing.

File: llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
+//===- ComputeModuleProperties.h - compute properties for module -===//
```
Comment (bader) [inline llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h:1]:
```suggestion
//===- ComputeModuleProperties.h - compute properties for module ----------===//
```
Usually, this part is exactly 80 characters long.

File: llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===--------------------------------------------------------------===//
+// Functions for computing module properties for SYCL modules.
+//===---------------------------------------------------------------===//
```
Comment (bader) [inline llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h:9]:
```suggestion
//===----------------------------------------------------------------------===//
// Functions for computing module properties for SYCL modules.
//===----------------------------------------------------------------------===//
```

**Review by bader — COMMENTED** [review]
LGTM, just minor comments.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 4 above)
Comment (sarnex) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: yeah i originally did it inverted to minimize requires changes to existing tests, but i agree it is more consistent. done in latest commit.

File: llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp (hunk 5 above)
Comment (sarnex) [reply llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp:10]: fixed in latest commit, thx

File: llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h (hunk 6 above)
Comment (sarnex) [reply llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h]: fixed in latest commit, thx

File: llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h (hunk 7 above)
Comment (sarnex) [reply llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h]: fixed in latest commit, thx

File: llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h [hunk 8]
```diff
@@ -0,0 +1,45 @@
+//===- ComputeModuleProperties.h - compute properties for module ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===---------------------------------------------------------------------===//
+// Functions for computing module properties for SYCL modules.
+//===---------------------------------------------------------------------===//
```
Comment (bader) [inline llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h:9]:
```suggestion
//===----------------------------------------------------------------------===//
// Functions for computing module properties for SYCL modules.
//===----------------------------------------------------------------------===//
```

**Review by bader — APPROVED** [review]
Thanks a lot for making this improvement!

File: llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h (hunk 8 above)
Comment (sarnex) [reply llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h]: thanks for the guidance! fixed in latest commit, thanks.

File: llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h [hunk 9]
```diff
@@ -0,0 +1,45 @@
+//===- ComputeModuleProperties.h - compute properties for module ----------===//
```
Comment (asudarsa) [inline llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h:1]: This file computes both module properties and symbols. May be the filename and description needs to include symbols as well? Another option will be to split and add a second file - ComputeModuleSymbols.cpp. 

Thanks

**Comment by asudarsa** [comment]
The change looks good to me overall. There is one concern that you may have considered already. Can you please mention about plans on changes to the wrapper logic if and when the properties and symbols files are not generated?

Thanks

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 4 above)
Comment (asudarsa) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: What does the -properties option accomplish? In the current implementation, I see that this option just controls the generation of .props file. is that accurate? Thanks

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 10]
```diff
@@ -201,6 +196,10 @@ cl::opt<module_split::IRSplitMode> SplitMode(
 cl::opt<bool> DoSymGen{"symbols", cl::desc("generate exported symbol files"),
                        cl::cat(PostLinkCat)};
 
+cl::opt<bool> DoPropGen{"properties",
```
Comment (asudarsa) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:199]: We have some checks in place for DoSymGen (towards the end of this file. Think we will need to add similar checks for this as well?

Thanks

File: llvm/test/tools/sycl-post-link/skip-properties-gen.ll [hunk 11]
```diff
@@ -0,0 +1,25 @@
+; This test verifies the behavior without the -properties option.
```
Comment (asudarsa) [inline llvm/test/tools/sycl-post-link/skip-properties-gen.ll:1]:
```suggestion
; This test verifies the behavior of sycl-post-link tool without the -properties and -symbols option.
```

File: llvm/test/tools/sycl-post-link/skip-properties-gen.ll [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+; In particular, we should see the properties and symbols file are not added to the output table.
```
Comment (asudarsa) [inline llvm/test/tools/sycl-post-link/skip-properties-gen.ll:2]:
```suggestion
; In particular, we should verify that the properties and symbols files are not added to the output table.
```

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 4 above)
Comment (asudarsa) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: I hate to suggest this (especially after Nick having modified all the tests to add -properties). I would have preferred '-skip-properties' and -skip-symbols' as they seem to be enabled by default and adding -properties and -symbols seems extraneous. But it's just a nit. 

Thanks

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 13]
```diff
@@ -10711,9 +10716,12 @@ static void getTripleBasedSYCLPostLinkOpts(const ToolChain &TC,
     bool SplitEsimd = TCArgs.hasFlag(
         options::OPT_fsycl_device_code_split_esimd,
         options::OPT_fno_sycl_device_code_split_esimd, SplitEsimdByDefault);
-    // Symbol file and specialization constant info generation is mandatory -
+    bool IsUsingLTO = TC.getDriver().isUsingLTO(/*IsDeviceOffloadAction=*/true);
```
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Clang.cpp:10719]: We will need to add this logic inside clang-linker-wrapper as well. Please see comment in Line 10671. I can help with it if need be.

Thanks

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 4 above)
Comment (bader) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: I think we are moving towards flipping the default. Once we move from `sycl-post-link` to `clang-linker-wrapper`, this option won't be needed.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 13 above)
Comment (sarnex) [reply clang/lib/Driver/ToolChains/Clang.cpp:10719]: I haven't added any thinLTO changes to clang-linker-wrapper yet. I will do that in my next change and will definitely include this. Thanks.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 4 above)
Comment (sarnex) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: agree, in the (near) future we will not need this options and can call these from `clang-linker-wrapper`.

**Comment by sarnex** [comment]
@asudarsa 
> Can you please mention about plans on changes to the wrapper logic if and when the properties and symbols files are not generated?

The plan is to call these functions directly from `clang-linker-wrapper`, both for thinLTO and normal linking

File: llvm/test/tools/sycl-post-link/skip-properties-gen.ll (hunk 12 above)
Comment (sarnex) [reply llvm/test/tools/sycl-post-link/skip-properties-gen.ll]: fixed in latest commit, thanks

File: llvm/test/tools/sycl-post-link/skip-properties-gen.ll (hunk 11 above)
Comment (sarnex) [reply llvm/test/tools/sycl-post-link/skip-properties-gen.ll]: fixed in latest commit, thanks

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 10 above)
Comment (sarnex) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:199]: good catch, fixed in latest commit, thank you!

File: llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h (hunk 9 above)
Comment (sarnex) [reply llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h:1]: both functions are basically guaranteed to always be used together so i'm not sure if splitting it into two files makes sense, and i couldn't think of a more descriptive name for a single file that isn't super long and over the character limit, so i'm planning on leaving this for now even if it isn't ideal. let me know if you prefer i address this now.

Comment (asudarsa) [reply llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h:1]: +1 for renaming if it's not much of a hassle. Will help someone new to quickly understand what we are doing. Thanks

**Review by asudarsa — APPROVED** [review]
LGTM to me as an intermediate change, Thanks

File: llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h (hunk 9 above)
Comment (sarnex) [reply llvm/include/llvm/SYCLLowerIR/ComputeModuleProperties.h]: renamed in latest commit, thanks

**Review by PietroGhg — APPROVED** [review]
Change in the Native CPU test LGTM, thank you

**Review by aelovikov-intel — APPROVED** [review]

**Review by maksimsab — APPROVED** [review]
LGTM.

nit: I would rather generate properties and symbols unconditionally since it doesn't throw errors and it is not really costly. Anyway, this is what we are going to do in clang-linker-wrapper.

**Comment by sarnex** [comment]
@maksimsab Thank you for the review!! I agree it's not costly, but it does incur extra file IO (which admittedly should not really be noticeable). The main idea here is that we can seperate out the splitting and props/symtable generation as seperate operations, which is important for thinLTO.

**Comment by sarnex** [comment]
@AlexeySachkov I know you're OOO but I'd like your review on this, I merged it now because there are a lot of changes and I had to rebase/merge almost daily. I'm happy to address any comments in a follow-up PR.

**Pull request merged**